### PR TITLE
Implement AcademyStats

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/academystats.cpp [hunk 1]
```diff
@@ -13,34 +13,761 @@
...
 
 void AcademyStats::Xfer_Snapshot(Xfer *xfer)
 {
-#ifdef GAME_DLL
-    Call_Method<void, SnapShot, Xfer *>(PICK_ADDRESS(0x00576BF0, 0x008DF4B4), this, xfer);
-#endif
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferUnsignedInt(&m_lastUpdateFrame);
+    xfer->xferBool(&m_needsUpdate);
```
Comment (xezon) [inline src/game/common/rts/academystats.cpp:731]: m_commandSet, m_commandCenterTemplate, m_supplyCenterTemplate are not serialized?

File: src/game/common/rts/academystats.cpp [hunk 2]
```diff
@@ -13,34 +13,761 @@
...
+    if (power->Get_Academy_Classify() == ACT_SUPERPOWER) {
+        m_supersUsedCount++;
+    }
+}
+
+void AcademyStats::Record_Income()
+{
+    unsigned int frame = g_theGameLogic->Get_Frame();
+
+    unsigned int income_frame = m_incomeFrame1 - frame;
```
Comment (xezon) [inline src/game/common/rts/academystats.cpp:281]: Looks different to `unsigned int power_frame = frame - m_powerFrame2;`. I wonder if there is mistake.

File: src/game/common/rts/academystats.cpp [hunk 3]
```diff
@@ -13,34 +13,761 @@
...
+    if (info == nullptr) {
+        captainslog_dbgassert(false, "AcademyStats::Calculate_Academy_Advice() was passed in NULL AcademyAdviceInfo.");
+        return false;
+    } else {
+        info->m_adviceCount = 0;
+
+        for (int i = 0; i < max_advice_count; i++) {
+            Utf16String str;
+            str = "\n\n";
+            info->m_adviceString[i].Format(str);
```
Comment (xezon) [inline src/game/common/rts/academystats.cpp:709]: This looks odd. It effectively just assigns \n\n to advice string?

File: src/game/common/rts/academystats.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/rts/academystats.cpp:731]: No they are not.

File: src/game/common/rts/academystats.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/common/rts/academystats.cpp:281]: This is correct.

File: src/game/common/rts/academystats.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/common/rts/academystats.cpp:709]: This is correct.

**Comment by xezon** [comment]
I do not know how to test this code.

**Comment by xezon** [comment]
Code can be tested by looking at "Tactical Analysis" text in Game Score screen. It shows different text presets depending on what the player has done in a match.

This works in principle.

**Review by xezon — APPROVED** [review]
Looks good.

**Pull request merged**